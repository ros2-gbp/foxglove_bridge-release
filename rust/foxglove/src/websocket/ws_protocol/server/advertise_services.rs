//! Advertise services message types.

use std::borrow::Cow;
use std::convert::TryFrom;

use serde::{Deserialize, Serialize};

use crate::websocket::ws_protocol::schema::{self, Schema};
use crate::websocket::ws_protocol::JsonMessage;

/// Advertise services message.
///
/// Spec: <https://github.com/foxglove/ws-protocol/blob/main/docs/spec.md#advertise-services>
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(tag = "op", rename = "advertiseServices", rename_all = "camelCase")]
pub struct AdvertiseServices<'a> {
    /// Services.
    #[serde(borrow)]
    pub services: Vec<Service<'a>>,
}

impl<'a> AdvertiseServices<'a> {
    /// Creates a new advertise services message.
    pub fn new(services: impl IntoIterator<Item = Service<'a>>) -> Self {
        Self {
            services: services.into_iter().collect(),
        }
    }

    /// Returns an owned version of this message.
    pub fn into_owned(self) -> AdvertiseServices<'static> {
        AdvertiseServices {
            services: self.services.into_iter().map(|s| s.into_owned()).collect(),
        }
    }
}

impl JsonMessage for AdvertiseServices<'_> {}

/// A service in a [`AdvertiseServices`] message.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Service<'a> {
    /// Service ID.
    pub id: u32,
    /// Service name.
    #[serde(borrow)]
    pub name: Cow<'a, str>,
    /// Service type, which may be used to derive the request & response schema.
    #[serde(borrow)]
    pub r#type: Cow<'a, str>,
    /// Request schema.
    #[serde(borrow, skip_serializing_if = "Option::is_none")]
    pub request: Option<MessageSchema<'a>>,
    /// Request schema name.
    #[serde(borrow, skip_serializing_if = "Option::is_none")]
    pub request_schema: Option<Cow<'a, str>>,
    /// Response schema.
    #[serde(borrow, skip_serializing_if = "Option::is_none")]
    pub response: Option<MessageSchema<'a>>,
    /// Response schema name.
    #[serde(borrow, skip_serializing_if = "Option::is_none")]
    pub response_schema: Option<Cow<'a, str>>,
}

impl<'a> Service<'a> {
    /// Creates a new service advertisement.
    pub fn new(id: u32, name: impl Into<Cow<'a, str>>, r#type: impl Into<Cow<'a, str>>) -> Self {
        Self {
            id,
            name: name.into(),
            r#type: r#type.into(),
            request: None,
            request_schema: Some("".into()),
            response: None,
            response_schema: Some("".into()),
        }
    }

    /// Adds a request schema to the service advertisement.
    ///
    /// This is the preferred method for setting request schemas. Note that calling
    /// [`Service::with_request_schema`] after this method will override the schema set here.
    pub fn with_request(
        mut self,
        encoding: impl Into<Cow<'a, str>>,
        schema: Schema<'a>,
    ) -> Result<Self, schema::EncodeError> {
        let schema_data = schema::encode_schema_data(&schema.encoding, schema.data)?;
        self.request = Some(MessageSchema::new(
            encoding,
            schema.name,
            schema.encoding,
            schema_data,
        ));
        self.request_schema = None;
        Ok(self)
    }

    /// Adds a response schema to the service advertisement.
    ///
    /// This is the preferred method for setting response schemas. Note that calling
    /// [`Service::with_response_schema`] after this method will override the schema set here.
    pub fn with_response(
        mut self,
        encoding: impl Into<Cow<'a, str>>,
        schema: Schema<'a>,
    ) -> Result<Self, schema::EncodeError> {
        let schema_data = schema::encode_schema_data(&schema.encoding, schema.data)?;
        self.response = Some(MessageSchema::new(
            encoding,
            schema.name,
            schema.encoding,
            schema_data,
        ));
        self.response_schema = None;
        Ok(self)
    }

    /// Adds a request schema name to the service advertisement.
    ///
    /// This method is provided for backwards compatibility. Prefer using [`Service::with_request`]
    /// instead. Note that this will override any schema set by [`Service::with_request`].
    #[must_use]
    pub fn with_request_schema(mut self, schema_name: impl Into<Cow<'a, str>>) -> Self {
        self.request = None;
        self.request_schema = Some(schema_name.into());
        self
    }

    /// Adds a response schema name to the service advertisement.
    ///
    /// This method is provided for backwards compatibility. Prefer using
    /// [`Service::with_response`] instead. Note that this will override any schema set by
    /// [`Service::with_response`].
    #[must_use]
    pub fn with_response_schema(mut self, schema_name: impl Into<Cow<'a, str>>) -> Self {
        self.response = None;
        self.response_schema = Some(schema_name.into());
        self
    }

    /// Returns an owned version of this service.
    pub fn into_owned(self) -> Service<'static> {
        Service {
            id: self.id,
            name: self.name.into_owned().into(),
            r#type: self.r#type.into_owned().into(),
            request: self.request.map(|r| r.into_owned()),
            request_schema: self.request_schema.map(|r| r.into_owned().into()),
            response: self.response.map(|r| r.into_owned()),
            response_schema: self.response_schema.map(|r| r.into_owned().into()),
        }
    }
}

/// A message schema for a [`Service`].
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MessageSchema<'a> {
    /// Message encoding.
    #[serde(borrow)]
    pub encoding: Cow<'a, str>,
    /// Schema name.
    #[serde(borrow)]
    pub schema_name: Cow<'a, str>,
    /// Schema encoding.
    #[serde(borrow)]
    pub schema_encoding: Cow<'a, str>,
    /// Schema data.
    ///
    /// This is the protocol-encoded form. You can use the [`MessageSchema::schema`] method to
    /// decode it.
    #[serde(borrow)]
    pub schema: Cow<'a, str>,
}

impl<'a> MessageSchema<'a> {
    /// Creates a new message schema.
    pub fn new(
        encoding: impl Into<Cow<'a, str>>,
        schema_name: impl Into<Cow<'a, str>>,
        schema_encoding: impl Into<Cow<'a, str>>,
        schema: impl Into<Cow<'a, str>>,
    ) -> Self {
        Self {
            encoding: encoding.into(),
            schema_name: schema_name.into(),
            schema_encoding: schema_encoding.into(),
            schema: schema.into(),
        }
    }

    /// Returns the decoded schema data.
    pub fn schema(&self) -> Result<Vec<u8>, schema::DecodeError> {
        schema::decode_schema_data(&self.schema_encoding, &self.schema)
    }

    /// Returns an owned version of this schema.
    pub fn into_owned(self) -> MessageSchema<'static> {
        MessageSchema {
            encoding: self.encoding.into_owned().into(),
            schema_name: self.schema_name.into_owned().into(),
            schema_encoding: self.schema_encoding.into_owned().into(),
            schema: Cow::Owned(self.schema.into_owned()),
        }
    }
}

impl<'a> TryFrom<MessageSchema<'a>> for Schema<'a> {
    type Error = schema::DecodeError;

    fn try_from(value: MessageSchema<'a>) -> Result<Self, Self::Error> {
        let schema = value.schema()?;
        Ok(Schema::new(
            value.schema_name,
            value.schema_encoding,
            schema,
        ))
    }
}

#[cfg(test)]
mod tests {
    use crate::websocket::ws_protocol::server::ServerMessage;

    use super::*;

    fn message() -> AdvertiseServices<'static> {
        AdvertiseServices::new([
            // Service with both request and response schemas using the new format
            Service::new(10, "/s1", "my_type")
                .with_request(
                    "json",
                    Schema::new("request-schema", "jsonschema", br#"{"type": "object"}"#),
                )
                .unwrap()
                .with_response(
                    "json",
                    Schema::new("response-schema", "jsonschema", br#"{"type": "object"}"#),
                )
                .unwrap(),
            // Service with only request schema using the new format
            Service::new(20, "/s2", "other_type")
                .with_request(
                    "protobuf",
                    Schema::new("request-schema", "protobuf", &[0xde, 0xad, 0xbe, 0xef]),
                )
                .unwrap(),
            // Service with both request and response schemas using the old format
            Service::new(30, "/s3", "old_type")
                .with_request_schema("request-schema")
                .with_response_schema("response-schema"),
            // Service with request schema in new format and response in old format
            Service::new(40, "/s4", "mixed_type")
                .with_request(
                    "json",
                    Schema::new("request-schema", "jsonschema", br#"{"type": "object"}"#),
                )
                .unwrap()
                .with_response_schema("response-schema"),
            // Service with request schema in old format and response in new format
            Service::new(50, "/s5", "mixed_type")
                .with_request_schema("request-schema")
                .with_response(
                    "json",
                    Schema::new("response-schema", "jsonschema", br#"{"type": "object"}"#),
                )
                .unwrap(),
            // Service with schema overrides
            Service::new(60, "/s6", "override_type")
                .with_request(
                    "json",
                    Schema::new("request-schema", "jsonschema", br#"{"type": "object"}"#),
                )
                .unwrap()
                .with_request_schema("new-request-schema")
                .with_response_schema("response-schema")
                .with_response(
                    "json",
                    Schema::new(
                        "new-response-schema",
                        "jsonschema",
                        br#"{"type": "object"}"#,
                    ),
                )
                .unwrap(),
            // Service with default schemas
            Service::new(70, "/s7", "default_schemas"),
        ])
    }

    #[test]
    fn test_encode() {
        insta::assert_json_snapshot!(message());
    }

    #[test]
    fn test_roundtrip() {
        let orig = message();
        let buf = orig.to_string();
        let msg = ServerMessage::parse_json(&buf).unwrap();
        assert_eq!(msg, ServerMessage::AdvertiseServices(orig));
    }
}
