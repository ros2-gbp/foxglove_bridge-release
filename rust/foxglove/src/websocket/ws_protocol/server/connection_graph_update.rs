//! Connection graph update message types.

use serde::{Deserialize, Serialize};

use crate::websocket::ws_protocol::JsonMessage;

/// Connection graph update message.
///
/// Spec: <https://github.com/foxglove/ws-protocol/blob/main/docs/spec.md#connection-graph-update>
#[derive(Debug, Default, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(tag = "op", rename = "connectionGraphUpdate", rename_all = "camelCase")]
pub struct ConnectionGraphUpdate {
    /// Published topics.
    pub published_topics: Vec<PublishedTopic>,
    /// Subscribed topics.
    pub subscribed_topics: Vec<SubscribedTopic>,
    /// Advertised services.
    pub advertised_services: Vec<AdvertisedService>,
    /// Removed tpoics.
    pub removed_topics: Vec<String>,
    /// Removed services.
    pub removed_services: Vec<String>,
}

impl JsonMessage for ConnectionGraphUpdate {}

/// A published topic in a [`ConnectionGraphUpdate`] message.
#[derive(Debug, Default, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PublishedTopic {
    /// Topic name.
    pub name: String,
    /// Topic subscriber IDs.
    pub publisher_ids: Vec<String>,
}

impl PublishedTopic {
    /// Creates a new published topic.
    pub fn new(
        name: impl Into<String>,
        publisher_ids: impl IntoIterator<Item = impl Into<String>>,
    ) -> Self {
        Self {
            name: name.into(),
            publisher_ids: publisher_ids.into_iter().map(|id| id.into()).collect(),
        }
    }
}

/// A subscribed topic in a [`ConnectionGraphUpdate`] message.
#[derive(Debug, Default, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SubscribedTopic {
    /// Topic name.
    pub name: String,
    /// Topic subscriber IDs.
    pub subscriber_ids: Vec<String>,
}

impl SubscribedTopic {
    /// Creates a new subscribed topic.
    pub fn new(
        name: impl Into<String>,
        subscriber_ids: impl IntoIterator<Item = impl Into<String>>,
    ) -> Self {
        Self {
            name: name.into(),
            subscriber_ids: subscriber_ids.into_iter().map(|id| id.into()).collect(),
        }
    }
}

/// An advertised service in a [`ConnectionGraphUpdate`] message.
#[derive(Debug, Default, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AdvertisedService {
    /// Service name.
    pub name: String,
    /// Service provider IDs.
    pub provider_ids: Vec<String>,
}

impl AdvertisedService {
    /// Creates a new advertised service.
    pub fn new(
        name: impl Into<String>,
        provider_ids: impl IntoIterator<Item = impl Into<String>>,
    ) -> Self {
        Self {
            name: name.into(),
            provider_ids: provider_ids.into_iter().map(|id| id.into()).collect(),
        }
    }
}

#[cfg(test)]
mod tests {
    use crate::websocket::ws_protocol::server::ServerMessage;

    use super::*;

    fn message() -> ConnectionGraphUpdate {
        ConnectionGraphUpdate {
            published_topics: vec![PublishedTopic::new("/t1", ["p1", "p2"])],
            subscribed_topics: vec![SubscribedTopic::new("/t2", ["s1", "s2"])],
            advertised_services: vec![AdvertisedService::new("/s1", ["pr1", "pr2"])],
            removed_topics: ["/t3", "/t4"].into_iter().map(String::from).collect(),
            removed_services: ["/s2", "/s3"].into_iter().map(String::from).collect(),
        }
    }

    #[test]
    fn test_encode() {
        insta::assert_json_snapshot!(message());
    }

    #[test]
    fn test_roundtrip() {
        let orig = message();
        let buf = orig.to_string();
        let msg = ServerMessage::parse_json(&buf).unwrap();
        assert_eq!(msg, ServerMessage::ConnectionGraphUpdate(orig));
    }
}
