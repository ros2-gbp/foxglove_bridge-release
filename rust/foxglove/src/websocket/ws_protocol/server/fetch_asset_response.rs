//! Fetch asset response types.

use std::borrow::Cow;

use bytes::{Buf, BufMut};

use crate::websocket::ws_protocol::{BinaryMessage, ParseError};

use super::BinaryOpcode;

/// Fetch asset response message.
///
/// Spec: <https://github.com/foxglove/ws-protocol/blob/main/docs/spec.md#fetch-asset-response>
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct FetchAssetResponse<'a> {
    /// Request ID.
    pub request_id: u32,
    /// Payload.
    pub payload: Payload<'a>,
}

impl<'a> FetchAssetResponse<'a> {
    /// Creates a new error message response.
    pub fn error_message(request_id: u32, message: impl Into<Cow<'a, str>>) -> Self {
        Self {
            request_id,
            payload: Payload::ErrorMessage(message.into()),
        }
    }

    /// Creates a new asset data response.
    pub fn asset_data(request_id: u32, data: impl Into<Cow<'a, [u8]>>) -> Self {
        Self {
            request_id,
            payload: Payload::AssetData(data.into()),
        }
    }

    /// Returns an owned version of this message.
    pub fn into_owned(self) -> FetchAssetResponse<'static> {
        FetchAssetResponse {
            request_id: self.request_id,
            payload: self.payload.into_owned(),
        }
    }
}

impl<'a> BinaryMessage<'a> for FetchAssetResponse<'a> {
    fn parse_binary(mut data: &'a [u8]) -> Result<Self, ParseError> {
        if data.len() < 4 + 1 + 4 {
            return Err(ParseError::BufferTooShort);
        }
        let request_id = data.get_u32_le();
        let raw_status = data.get_u8();
        let Some(status) = Status::from_repr(raw_status) else {
            return Err(ParseError::InvalidFetchAssetStatus(raw_status));
        };
        let error_message_len = data.get_u32_le() as usize;
        if data.len() < error_message_len {
            return Err(ParseError::BufferTooShort);
        }
        let payload = match status {
            Status::Success => {
                data.advance(error_message_len);
                Payload::AssetData(Cow::Borrowed(data))
            }
            Status::Error => {
                let msg = std::str::from_utf8(&data[..error_message_len])?;
                Payload::ErrorMessage(Cow::Borrowed(msg))
            }
        };
        Ok(Self {
            request_id,
            payload,
        })
    }

    fn to_bytes(&self) -> Vec<u8> {
        let (status, error_message_len, payload_len, payload) = match &self.payload {
            Payload::ErrorMessage(msg) => (Status::Error, msg.len(), msg.len(), msg.as_bytes()),
            Payload::AssetData(data) => (Status::Success, 0, data.len(), data.as_ref()),
        };
        let size = 1 + 4 + 1 + 4 + payload_len;
        let mut buf = Vec::with_capacity(size);
        buf.put_u8(BinaryOpcode::FetchAssetResponse as u8);
        buf.put_u32_le(self.request_id);
        buf.put_u8(status as u8);
        buf.put_u32_le(error_message_len as u32);
        buf.put_slice(payload);
        buf
    }
}

/// Status code.
#[repr(u8)]
enum Status {
    Success = 0,
    Error = 1,
}
impl Status {
    fn from_repr(value: u8) -> Option<Self> {
        match value {
            0 => Some(Self::Success),
            1 => Some(Self::Error),
            _ => None,
        }
    }
}

/// Payload for a [`FetchAssetResponse`].
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Payload<'a> {
    /// An error response.
    ErrorMessage(Cow<'a, str>),
    /// A successful response.
    AssetData(Cow<'a, [u8]>),
}

impl Payload<'_> {
    /// Returns an owned version of this payload.
    pub fn into_owned(self) -> Payload<'static> {
        match self {
            Payload::ErrorMessage(msg) => Payload::ErrorMessage(Cow::Owned(msg.into_owned())),
            Payload::AssetData(data) => Payload::AssetData(Cow::Owned(data.into_owned())),
        }
    }
}

#[cfg(test)]
mod tests {
    use assert_matches::assert_matches;

    use crate::websocket::ws_protocol::server::ServerMessage;

    use super::*;

    fn asset_data() -> FetchAssetResponse<'static> {
        FetchAssetResponse::asset_data(10, b"data")
    }

    fn error_message() -> FetchAssetResponse<'static> {
        FetchAssetResponse::error_message(10, "oh no")
    }

    #[test]
    fn test_encode_asset_data() {
        insta::assert_snapshot!(format!("{:#04x?}", asset_data().to_bytes()));
    }

    #[test]
    fn test_encode_error_message() {
        insta::assert_snapshot!(format!("{:#04x?}", error_message().to_bytes()));
    }

    #[test]
    fn test_parse() {
        assert_matches!(
            FetchAssetResponse::parse_binary(b""),
            Err(ParseError::BufferTooShort)
        );
        assert_matches!(
            FetchAssetResponse::parse_binary(&[0; 8]),
            Err(ParseError::BufferTooShort)
        );

        let mut buf = Vec::new();
        buf.put_u32_le(10);
        buf.put_u8(2);
        buf.put_u32_le(0);
        assert_matches!(
            FetchAssetResponse::parse_binary(&buf),
            Err(ParseError::InvalidFetchAssetStatus(2))
        );

        let mut buf = Vec::new();
        buf.put_u32_le(10);
        buf.put_u8(1);
        buf.put_u32_le(1);
        assert_matches!(
            FetchAssetResponse::parse_binary(&buf),
            Err(ParseError::BufferTooShort)
        );
    }

    #[test]
    fn test_roundtrip_asset_data() {
        let orig = asset_data();
        let buf = orig.to_bytes();
        let msg = ServerMessage::parse_binary(&buf).unwrap();
        assert_eq!(msg, ServerMessage::FetchAssetResponse(orig));
    }

    #[test]
    fn test_roundtrip_error_message() {
        let orig = error_message();
        let buf = orig.to_bytes();
        let msg = ServerMessage::parse_binary(&buf).unwrap();
        assert_eq!(msg, ServerMessage::FetchAssetResponse(orig));
    }
}
