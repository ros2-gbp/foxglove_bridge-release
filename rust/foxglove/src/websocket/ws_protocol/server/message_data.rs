use std::borrow::Cow;

use bytes::{Buf, BufMut};

use crate::websocket::ws_protocol::{BinaryMessage, ParseError};

use super::BinaryOpcode;

/// Message data message.
///
/// Spec: <https://github.com/foxglove/ws-protocol/blob/main/docs/spec.md#message-data>
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct MessageData<'a> {
    /// Subscription ID.
    pub subscription_id: u32,
    /// Log time.
    pub log_time: u64,
    /// Message data.
    pub data: Cow<'a, [u8]>,
}

impl<'a> MessageData<'a> {
    /// Creates a new message data message.
    pub fn new(subscription_id: u32, log_time: u64, data: impl Into<Cow<'a, [u8]>>) -> Self {
        Self {
            subscription_id,
            log_time,
            data: data.into(),
        }
    }

    /// Returns an owned version of this message.
    pub fn into_owned(self) -> MessageData<'static> {
        MessageData {
            subscription_id: self.subscription_id,
            log_time: self.log_time,
            data: Cow::Owned(self.data.into_owned()),
        }
    }
}

impl<'a> BinaryMessage<'a> for MessageData<'a> {
    fn parse_binary(mut data: &'a [u8]) -> Result<Self, ParseError> {
        if data.len() < 4 + 8 {
            return Err(ParseError::BufferTooShort);
        }
        let subscription_id = data.get_u32_le();
        let log_time = data.get_u64_le();
        Ok(Self {
            subscription_id,
            log_time,
            data: Cow::Borrowed(data),
        })
    }

    fn to_bytes(&self) -> Vec<u8> {
        let size = 1 + 4 + 8 + self.data.len();
        let mut buf = Vec::with_capacity(size);
        buf.put_u8(BinaryOpcode::MessageData as u8);
        buf.put_u32_le(self.subscription_id);
        buf.put_u64_le(self.log_time);
        buf.put_slice(&self.data);
        buf
    }
}

#[cfg(test)]
mod tests {
    use crate::websocket::ws_protocol::server::ServerMessage;

    use super::*;

    fn message() -> MessageData<'static> {
        MessageData {
            subscription_id: 30,
            log_time: 1234,
            data: br#"{"key": "value"}"#.into(),
        }
    }

    #[test]
    fn test_encode() {
        insta::assert_snapshot!(format!("{:#04x?}", message().to_bytes()));
    }

    #[test]
    fn test_roundtrip() {
        let orig = message();
        let buf = orig.to_bytes();
        let msg = ServerMessage::parse_binary(&buf).unwrap();
        assert_eq!(msg, ServerMessage::MessageData(orig));
    }
}
