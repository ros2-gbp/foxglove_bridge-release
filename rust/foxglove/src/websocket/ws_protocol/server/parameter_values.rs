use serde::{Deserialize, Serialize};

use crate::websocket::ws_protocol::parameter::Parameter;
use crate::websocket::ws_protocol::JsonMessage;

/// Parameter values message.
///
/// Spec: <https://github.com/foxglove/ws-protocol/blob/main/docs/spec.md#parameter-values>
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(tag = "op", rename = "parameterValues", rename_all = "camelCase")]
pub struct ParameterValues {
    /// Parameter values.
    pub parameters: Vec<Parameter>,
    /// ID from a get/set parameters request.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub id: Option<String>,
}

impl ParameterValues {
    /// Creates a new parameter values message.
    pub fn new(parameters: impl IntoIterator<Item = Parameter>) -> Self {
        Self {
            parameters: parameters.into_iter().collect(),
            id: None,
        }
    }

    /// Sets the request ID.
    #[must_use]
    pub fn with_id(mut self, id: impl Into<String>) -> Self {
        self.id = Some(id.into());
        self
    }
}

impl JsonMessage for ParameterValues {}

#[cfg(test)]
mod tests {
    use crate::websocket::ws_protocol::server::ServerMessage;

    use super::*;

    fn message() -> ParameterValues {
        ParameterValues::new([
            Parameter::empty("empty"),
            Parameter::float64("f64", 1.23),
            Parameter::float64_array("f64[]", [1.23, 4.56]),
            Parameter::string("str", "hello"),
            Parameter::byte_array("byte[]", &[0x10, 0x20, 0x30]),
            Parameter::bool("bool", true),
        ])
    }

    fn message_with_id() -> ParameterValues {
        message().with_id("my-id")
    }

    #[test]
    fn test_encode() {
        insta::assert_json_snapshot!(message());
    }

    #[test]
    fn test_encode_with_id() {
        insta::assert_json_snapshot!(message_with_id());
    }

    fn test_roundtrip_inner(orig: ParameterValues) {
        let buf = orig.to_string();
        let msg = ServerMessage::parse_json(&buf).unwrap();
        assert_eq!(msg, ServerMessage::ParameterValues(orig));
    }

    #[test]
    fn test_roundtrip() {
        test_roundtrip_inner(message())
    }

    #[test]
    fn test_roundtrip_with_id() {
        test_roundtrip_inner(message_with_id())
    }
}
