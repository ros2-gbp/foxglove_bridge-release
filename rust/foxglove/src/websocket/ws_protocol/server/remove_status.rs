use serde::{Deserialize, Serialize};

use crate::websocket::ws_protocol::JsonMessage;

/// Remove status message.
///
/// Spec: <https://github.com/foxglove/ws-protocol/blob/main/docs/spec.md#remove-status>
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(tag = "op", rename = "removeStatus", rename_all = "camelCase")]
pub struct RemoveStatus {
    /// IDs of the status messages to be removed.
    pub status_ids: Vec<String>,
}

impl RemoveStatus {
    /// Creates a new remove status message.
    pub fn new(status_ids: impl IntoIterator<Item = impl Into<String>>) -> Self {
        Self {
            status_ids: status_ids.into_iter().map(|s| s.into()).collect(),
        }
    }
}

impl JsonMessage for RemoveStatus {}

#[cfg(test)]
mod tests {
    use crate::websocket::ws_protocol::server::ServerMessage;

    use super::*;

    fn message() -> RemoveStatus {
        RemoveStatus::new(["status-1", "status-2", "status-3"])
    }

    #[test]
    fn test_encode() {
        insta::assert_json_snapshot!(message());
    }

    #[test]
    fn test_roundtrip() {
        let orig = message();
        let buf = orig.to_string();
        let msg = ServerMessage::parse_json(&buf).unwrap();
        assert_eq!(msg, ServerMessage::RemoveStatus(orig));
    }
}
