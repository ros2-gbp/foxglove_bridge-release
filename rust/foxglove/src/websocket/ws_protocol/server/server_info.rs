//! Server info message types.

use std::collections::HashMap;

use serde::{Deserialize, Serialize};

use crate::websocket::ws_protocol::JsonMessage;

/// Server info message.
///
/// Spec: <https://github.com/foxglove/ws-protocol/blob/main/docs/spec.md#server-info>
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(tag = "op", rename = "serverInfo", rename_all = "camelCase")]
pub struct ServerInfo {
    /// Free-form information about the server.
    pub name: String,
    /// The optional features supported by this server.
    pub capabilities: Vec<Capability>,
    /// The encodings that may be used for client-side publishing or service call
    /// requests/responses. Only set if client publishing or services are supported.
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub supported_encodings: Vec<String>,
    /// Optional map of key-value pairs.
    #[serde(default, skip_serializing_if = "HashMap::is_empty")]
    pub metadata: HashMap<String, String>,
    /// Optional string.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub session_id: Option<String>,
}
impl ServerInfo {
    /// Creates a new server info message.
    pub fn new(name: impl Into<String>) -> Self {
        Self {
            name: name.into(),
            capabilities: vec![],
            supported_encodings: vec![],
            metadata: HashMap::default(),
            session_id: None,
        }
    }

    /// Sets advertised capabilities.
    #[must_use]
    pub fn with_capabilities(mut self, capabilities: impl IntoIterator<Item = Capability>) -> Self {
        self.capabilities = capabilities.into_iter().collect();
        self
    }

    /// Sets supported encodings.
    #[must_use]
    pub fn with_supported_encodings(
        mut self,
        encodings: impl IntoIterator<Item = impl Into<String>>,
    ) -> Self {
        self.supported_encodings = encodings.into_iter().map(|s| s.into()).collect();
        self
    }

    /// Sets metadata.
    #[must_use]
    pub fn with_metadata(mut self, metadata: HashMap<String, String>) -> Self {
        self.metadata = metadata;
        self
    }

    /// Sets session ID.
    #[must_use]
    pub fn with_session_id(mut self, session_id: impl Into<String>) -> Self {
        self.session_id = Some(session_id.into());
        self
    }
}

impl JsonMessage for ServerInfo {}

/// A capability advertised in a [`ServerInfo`] message.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum Capability {
    /// Allow clients to advertise channels to send data messages to the server.
    ClientPublish,
    /// Allow clients to get & set parameters.
    Parameters,
    /// Allow clients to subscribe to parameter changes.
    ParametersSubscribe,
    /// The server may publish binary time messages.
    Time,
    /// Allow clients to call services.
    Services,
    /// Allow clients to subscribe to updates to the connection graph.
    ConnectionGraph,
    /// Allow clients to fetch assets.
    Assets,
}

#[cfg(test)]
mod tests {
    use crate::websocket::ws_protocol::server::ServerMessage;

    use super::*;

    fn message() -> ServerInfo {
        ServerInfo::new("example server")
    }

    fn message_full() -> ServerInfo {
        message()
            .with_capabilities([Capability::ClientPublish, Capability::Time])
            .with_supported_encodings(["json"])
            .with_metadata(maplit::hashmap! {
                "key".into() => "value".into(),
            })
            .with_session_id("1675789422160")
    }

    #[test]
    fn test_encode() {
        insta::assert_json_snapshot!(message());
    }

    #[test]
    fn test_encode_full() {
        insta::assert_json_snapshot!(message_full());
    }

    fn test_roundtrip_inner(orig: ServerInfo) {
        let buf = orig.to_string();
        let msg = ServerMessage::parse_json(&buf).unwrap();
        assert_eq!(msg, ServerMessage::ServerInfo(orig));
    }

    #[test]
    fn test_roundtrip() {
        test_roundtrip_inner(message());
    }

    #[test]
    fn test_roundtrip_full() {
        test_roundtrip_inner(message_full());
    }
}
