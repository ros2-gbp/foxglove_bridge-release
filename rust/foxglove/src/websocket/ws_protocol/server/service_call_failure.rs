use serde::{Deserialize, Serialize};

use crate::websocket::ws_protocol::JsonMessage;

/// Service call failure message.
///
/// Spec: <https://github.com/foxglove/ws-protocol/blob/main/docs/spec.md#service-call-failure>
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(tag = "op", rename = "serviceCallFailure", rename_all = "camelCase")]
pub struct ServiceCallFailure {
    /// Service ID.
    pub service_id: u32,
    /// Call ID.
    pub call_id: u32,
    /// Error message.
    pub message: String,
}

impl ServiceCallFailure {
    /// Creates a new service call failure message.
    pub fn new(service_id: u32, call_id: u32, message: impl Into<String>) -> Self {
        Self {
            service_id,
            call_id,
            message: message.into(),
        }
    }
}

impl JsonMessage for ServiceCallFailure {}

#[cfg(test)]
mod tests {
    use crate::websocket::ws_protocol::server::ServerMessage;

    use super::*;

    fn message() -> ServiceCallFailure {
        ServiceCallFailure {
            service_id: 1,
            call_id: 1,
            message: "Service does not exist".into(),
        }
    }

    #[test]
    fn test_encode() {
        insta::assert_json_snapshot!(message());
    }

    #[test]
    fn test_roundtrip() {
        let orig = message();
        let buf = orig.to_string();
        let msg = ServerMessage::parse_json(&buf).unwrap();
        assert_eq!(msg, ServerMessage::ServiceCallFailure(orig));
    }
}
