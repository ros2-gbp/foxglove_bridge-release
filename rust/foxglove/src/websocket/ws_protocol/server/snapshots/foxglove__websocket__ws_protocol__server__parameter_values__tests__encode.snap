---
source: rust/foxglove/src/websocket/ws_protocol/server/parameter_values.rs
expression: message()
---
{
  "op": "parameterValues",
  "parameters": [
    {
      "name": "empty"
    },
    {
      "name": "f64",
      "type": "float64",
      "value": 1.23
    },
    {
      "name": "f64[]",
      "type": "float64_array",
      "value": [
        1.23,
        4.56
      ]
    },
    {
      "name": "str",
      "value": "hello"
    },
    {
      "name": "byte[]",
      "type": "byte_array",
      "value": "ECAw"
    },
    {
      "name": "bool",
      "value": true
    }
  ]
}
